### PR TITLE
[FIX] mail: hide message failures for deleted records

### PR DESCRIPTION
**Pull request by yajo**
Description of the issue/feature this PR addresses:
Do not display useless notification failures.

Current behavior before PR:
When a record that inherits from `mail.thread` is deleted by an `ondelete="cascade"` SQL constraint, its messages and notifiations are kept in DB.

If some of those notifications is in status `bounce` or `exception`, it will keep being displayed to the user. However, the user won't be able to do anything with it. Actually, if they try to, they'll get an exception.

<details>
<summary>Exception</summary>

```
Traceback (most recent call last):
  File \"/opt/odoo/custom/src/odoo/odoo/api.py\", line 997, in get
    cache_value = field_cache[record._ids[0]]
KeyError: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1161, in __get__
    value = env.cache.get(record, self)
  File \"/opt/odoo/custom/src/odoo/odoo/api.py\", line 1004, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'account.payment(1,).company_id'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"/opt/odoo/custom/src/odoo/odoo/api.py\", line 997, in get
    cache_value = field_cache[record._ids[0]]
KeyError: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1161, in __get__
    value = env.cache.get(record, self)
  File \"/opt/odoo/custom/src/odoo/odoo/api.py\", line 1004, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'account.payment(1,).move_id'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1210, in __get__
    self.compute_value(recs)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1392, in compute_value
    records._compute_field_value(self)
  File \"/opt/odoo/auto/addons/mail/models/mail_thread.py\", line 403, in _compute_field_value
    return super()._compute_field_value(field)
  File \"/opt/odoo/custom/src/odoo/odoo/models.py\", line 4241, in _compute_field_value
    fields.determine(field.compute, self)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 101, in determine
    return needle(records, *args)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 690, in _compute_related
    values = [first(value[name]) for value in values]
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 690, in <listcomp>
    values = [first(value[name]) for value in values]
  File \"/opt/odoo/custom/src/odoo/odoo/models.py\", line 5957, in __getitem__
    return self._fields[key].__get__(self, self.env.registry[self._name])
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 2804, in __get__
    return super().__get__(records, owner)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1191, in __get__
    raise MissingError(\"\
\".join([
odoo.exceptions.MissingError: Record does not exist or has been deleted.
(Record: account.payment(1,), User: 2)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"/opt/odoo/custom/src/odoo/odoo/api.py\", line 997, in get
    cache_value = field_cache[record._ids[0]]
KeyError: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1161, in __get__
    value = env.cache.get(record, self)
  File \"/opt/odoo/custom/src/odoo/odoo/api.py\", line 1004, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'account.payment(1,).move_id'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"/opt/odoo/custom/src/odoo/odoo/http.py\", line 1652, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File \"/opt/odoo/custom/src/odoo/odoo/service/model.py\", line 133, in retrying
    result = func()
  File \"/opt/odoo/custom/src/odoo/odoo/http.py\", line 1679, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File \"/opt/odoo/custom/src/odoo/odoo/http.py\", line 1883, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File \"/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py\", line 154, in _dispatch
    result = endpoint(**request.params)
  File \"/opt/odoo/custom/src/odoo/odoo/http.py\", line 734, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File \"/opt/odoo/auto/addons/web/controllers/dataset.py\", line 42, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File \"/opt/odoo/auto/addons/web/controllers/dataset.py\", line 33, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File \"/opt/odoo/custom/src/odoo/odoo/api.py\", line 464, in call_kw
    result = _call_kw_model(method, model, args, kwargs)
  File \"/opt/odoo/custom/src/odoo/odoo/api.py\", line 435, in _call_kw_model
    result = method(recs, *args, **kwargs)
  File \"/opt/odoo/auto/addons/snailmail/models/mail_thread.py\", line 22, in notify_cancel_by_type
    super().notify_cancel_by_type(notification_type)
  File \"/opt/odoo/auto/addons/sms/models/mail_thread.py\", line 383, in notify_cancel_by_type
    super().notify_cancel_by_type(notification_type)
  File \"/opt/odoo/auto/addons/mail/models/mail_thread.py\", line 2420, in notify_cancel_by_type
    self._notify_cancel_by_type_generic('email')
  File \"/opt/odoo/auto/addons/mail/models/mail_thread.py\", line 2406, in _notify_cancel_by_type_generic
    self.env['mail.message'].browse(msg_ids)._notify_message_notification_update()
  File \"/opt/odoo/auto/addons/mail/models/mail_message.py\", line 1017, in _notify_message_notification_update
    record.check_access_rule('read')
  File \"/opt/odoo/custom/src/odoo/odoo/models.py\", line 3494, in check_access_rule
    invalid = self - self._filter_access_rules_python(operation)
  File \"/opt/odoo/custom/src/odoo/odoo/models.py\", line 3549, in _filter_access_rules_python
    return self.sudo().filtered_domain(dom or [])
  File \"/opt/odoo/custom/src/odoo/odoo/models.py\", line 5508, in filtered_domain
    data = record.mapped(key)
  File \"/opt/odoo/custom/src/odoo/odoo/models.py\", line 5428, in mapped
    recs = recs._fields[name].mapped(recs)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1284, in mapped
    self.__get__(first(remaining), type(remaining))
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 2804, in __get__
    return super().__get__(records, owner)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1212, in __get__
    self.compute_value(record)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1392, in compute_value
    records._compute_field_value(self)
  File \"/opt/odoo/auto/addons/mail/models/mail_thread.py\", line 403, in _compute_field_value
    return super()._compute_field_value(field)
  File \"/opt/odoo/custom/src/odoo/odoo/models.py\", line 4241, in _compute_field_value
    fields.determine(field.compute, self)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 101, in determine
    return needle(records, *args)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 690, in _compute_related
    values = [first(value[name]) for value in values]
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 690, in <listcomp>
    values = [first(value[name]) for value in values]
  File \"/opt/odoo/custom/src/odoo/odoo/models.py\", line 5957, in __getitem__
    return self._fields[key].__get__(self, self.env.registry[self._name])
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 2804, in __get__
    return super().__get__(records, owner)
  File \"/opt/odoo/custom/src/odoo/odoo/fields.py\", line 1191, in __get__
    raise MissingError(\"\
\".join([
odoo.exceptions.MissingError: Record does not exist or has been deleted.
(Record: account.payment(1,), User: 2)
```

</details>

More details:
https://www.loom.com/share/a59e9f0bfb754711a970608247433459?sid=e97a41e1-7615-45ff-a6fc-086507d08130

Desired behavior after PR is merged:

This fix avoids displaying failures related to non-existing records. This clears noise and avoids situations where the user is unable to react to their notifications.

@moduon MT-7609 OPW-4285691

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by seb-odoo** [comment]
Hello,

I'm not sure what is @tde-banana-odoo stance on this, but I believe the issue is to rely on cascade in this flow.

We specifically have an override of unlink in thread to unlink the related messages.

The problematic flow should be adapted to ensure the unlink of each thread is properly called.

I'm not fully aware of accounting code, but in this case, unlink of journal should probably call unlink of related payments, or something like that.

We definitely don't want to post filter notifications in `_message_fetch_failed` and let them live in a dead state in database.

If we must live with them, then we should adapt the code to ensure they can be marked as read (or deleted) and at least ensure there is no crash when the related record is deleted.

**Comment by yajo** [comment]
> The problematic flow should be adapted to ensure the unlink of each thread is properly called.

I'm not sure if that's technically possible. It affects any model, as long as it inherits `mail.thread` and has some field with `ondelete="cascade"`. Our customer just happened to trigger it in account payments, but the issue is at the DB schema level.

In those cases, Postgres will remove the records without telling the ORM, which in turn will have no way to know that there are pending unlinks to call.

There are other places where this can happen, like `ir.config_parameter`. This table, just like `mail.message`, holds a relaxed reference to other DB records. Instead of adding a simple FK constraint, they hold the model name and res_id separately, and hope for the best.

A Postgres trigger could be added to those tables, so the cascades actually reach them and there's no garbage left in the database. However, that could only be added to master due to DB schema immutability policies in stable versions.

In summary, you probably hold already thousands of dangling records in any database. Cascade deletions are everywhere, and relaxed references too. I thought that this would be the least dramatic fix. But I agree a better fix is desirable for master.

**Comment by tde-banana-odoo** [comment]
Indeed ondelete bypasses everything, that's really annoying. We could imagine a garbage collect job that tries to cleanup leftover of records, but I don't know how this could be efficient as we have a constantly growing history of messages, most of them being legit.

For stable, couldn't we improve "cancel" code so that if checks for record existence and remove notifications if not found, instead of crashing ? As we don't use this wizard in (big) batch, it is not that annoying if it checks for record existence. That way at least you can cleanup notifications instead of having them for eternity.

For master a real cascade would be interesting, I agree.

**Comment by yajo** [comment]
Another possibility we have is to add automated deletion of exception and bounce notifications after 6 months. It would mean changing a bit this domain: https://github.com/odoo/odoo/blob/a7edf019e67f84764b23e0b4e699adc5100aeb5f/addons/mail/models/mail_notification.py#L90-L96

What would you prefer?

**Comment by tde-banana-odoo** [comment]
> Another possibility we have is to add automated deletion of exception and bounce notifications after 6 months. It would mean changing a bit this domain:
> 
> https://github.com/odoo/odoo/blob/a7edf019e67f84764b23e0b4e699adc5100aeb5f/addons/mail/models/mail_notification.py#L90-L96
> 
> What would you prefer?

It is limited anyway to internal users, to keep trace of external customers that were notified (e.g. if someone says "I never received your invoice 8 months ago, I won't pay"). So it will not solve issues for those partners :) .

**Comment by yajo** [comment]
> couldn't we improve "cancel" code so that if checks for record existence and remove notifications if not found, instead of crashing ?

As you see in the video above, it already crashes when trying to _see_ notifications, and also when trying to _dismiss_ them.

With the current fix, there are no crashes. But would you prefer to have the 1st crash as long as we let the users dismiss the failure (fixing only the 2nd)?



> if someone says "I never received your invoice 8 months ago, I won't pay"

Then we still want to keep those send failures in DB? Then isn't the current fix better because it won't remove anything? It just won't annoy the users with non-actionable notification bubbles that raise errors. But you can still find those failures in DB if you need them later.

**Comment by yajo** [comment]
FYI Here you have a new, slightly related fix: https://github.com/odoo/odoo/pull/206829. It's another side-effect from the same problem (FK removals). That one is more dangerous because it blocks the outgoing mail cron.

File: addons/mail/models/res_partner.py [hunk 1]
```diff
@@ -141,6 +142,14 @@ def _message_fetch_failed(self):
...
             ('mail_message_id.res_id', '!=', 0),
         ], limit=100)
+        found = defaultdict(list)
+        for message in notifications.mail_message_id:
+            found[message.model].append(message.res_id)
+        existing = {
+            model: set(self.env[model].browse(ids).exists().ids)
+            for model, ids in found.items()
+        }
+        notifications = notifications.filtered(lambda n: n.mail_message_id.res_id in existing[n.mail_message_id.model])
```
Comment (seb-odoo) [inline addons/mail/models/res_partner.py:152]: So shouldn't we actually change the `notification_status` of the invalid notifications on the fly (to `canceled` state) so they don't need to be fetched and filtered again the next time?
Edit: or just deleted them, I'm not sure.

Because if not doing so, if there are more than 100 of those, it would lead to a never resolvable state of not returning any notification ever again as they will always all be filtered out.

Comment (seb-odoo) [reply addons/mail/models/res_partner.py:152]: Could you add a test to cover the current fix + the potential extra fix of changing their state?
